Gradle sync fail - ERROR: Text must not be null or empty

## Env. Info
```
System:
    OS: macOS Mojave 10.14.4
    CPU: (8) x64 Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz
    Memory: 1.20 GB / 16.00 GB
    Shell: 5.3 - /bin/zsh
  Binaries:
    Node: 10.13.0 - ~/.nvm/versions/node/v10.13.0/bin/node
    Yarn: 1.12.3 - /usr/local/bin/yarn
    npm: 6.10.2 - ~/.nvm/versions/node/v10.13.0/bin/npm
    Watchman: 4.9.0 - /usr/local/bin/watchman
  SDKs:
    iOS SDK:
      Platforms: iOS 12.4, macOS 10.14, tvOS 12.4, watchOS 5.3
    Android SDK:
      API Levels: 26, 28, 29
      Build Tools: 28.0.3, 29.0.2
      System Images: android-29 | Google APIs Intel x86 Atom
  IDEs:
    Android Studio: 3.5 AI-191.8026.42.35.5900203
    Xcode: 10.3/10G8 - /usr/bin/xcodebuild
  npmPackages:
    react: 16.9.0 => 16.9.0 
    react-native: 0.61.2 => 0.61.2
```



## Current cli version.

 Andrei  ~  desktop  ReactConf   native_navigation ✚ 1 … 1  yarn list @react-native-community/cli 
yarn list v1.12.3
warning Filtering by arguments is deprecated. Please use the pattern option instead.
└─ @react-native-community/cli@3.0.0-alpha.2
✨  Done in 0.57s.


## Description

I'm able to build and run on device using `npx react-native run-android`, however, when I open Android Studio I'm unable to sync gradle. 

## Reproducible Demo
Clean init with `npx react-native init FreshApp`

```  
Caused by: org.gradle.internal.exceptions.LocationAwareException: Script '/Users/Andrei/desktop/ReactConf/node_modules/@react-native-community/cli-platform-android/native_modules.gradle' line: 191
A problem occurred evaluating settings 'ReactConf'.

```

Full log file from Gradle failure:

[idea.log](https://github.com/react-native-community/cli/files/3729600/idea.log)




## Note

I noticed some other issues related to this, I tried to going over them and to find a fix, but I failed so far. Most seem related to a Monorepo structure, which is not my case. 


Any clues how I can fix this?

--- a/packages/cli/src/tools/config/resolveReactNativePath.ts
+++ b/packages/cli/src/tools/config/resolveReactNativePath.ts
@@ -11,16 +11,14 @@ export default function resolveReactNativePath(root: string) {
     return resolveNodeModuleDir(root, 'react-native');
   } catch (_ignored) {
     throw new CLIError(`
-      Unable to find React Native files. Make sure "react-native" module is installed
+      Unable to find React Native files looking up from ${root}. Make sure "react-native" module is installed
       in your project dependencies.
 
       If you are using React Native from a non-standard location, consider setting:
       {
-        "react-native": {
-          "reactNativePath": "./path/to/react-native"
-        }
+        reactNativePath: "./path/to/react-native"
       }
-      in your \`package.json\`.
+      in your \`react-native.config.js\`.
     `);
   }
 }

--- a/packages/platform-android/native_modules.gradle
+++ b/packages/platform-android/native_modules.gradle
@@ -2,6 +2,7 @@ import groovy.json.JsonSlurper
 import org.gradle.initialization.DefaultSettings
 import org.apache.tools.ant.taskdefs.condition.Os
 
+def jsAppDir = buildscript.getProperties().get("sourceFile").toString().split("node_modules/@react-native-community/cli-platform-android/native_modules.gradle")[0]
 def generatedFileName = "PackageList.java"
 def generatedFilePackage = "com.facebook.react"
 def generatedFileContentsTemplate = """
@@ -71,12 +72,14 @@ public class PackageList {
 class ReactNativeModules {
   private Logger logger
   private String packageName
+  private String jsAppDir
   private ArrayList<HashMap<String, String>> reactNativeModules
 
   private static String LOG_PREFIX = ":ReactNative:"
 
-  ReactNativeModules(Logger logger) {
+  ReactNativeModules(Logger logger, String jsAppDir) {
     this.logger = logger
+    this.jsAppDir = jsAppDir
 
     def (nativeModules, packageName) = this.getReactNativeConfig()
     this.reactNativeModules = nativeModules
@@ -150,13 +153,13 @@ class ReactNativeModules {
     if (this.reactNativeModules != null) return this.reactNativeModules
     ArrayList<HashMap<String, String>> reactNativeModules = new ArrayList<HashMap<String, String>>()
 
-    def cmdProcess
     def npx = Os.isFamily(Os.FAMILY_WINDOWS) ? "npx.cmd" : "npx"
-    def command = "${npx} --quiet react-native config"
+    def command = "${npx} --quiet --no-install react-native config"
     def reactNativeConfigOutput = ""
 
     try {
-      cmdProcess = Runtime.getRuntime().exec(command)
+      // When running Gradle from CWD that's outside of JS project, it's necessary to pass a CWD that's close to JS app,// so that npx can resolve to correct local CLI, instead of possibly installed global, corrupting config output
+      def cmdProcess = Runtime.getRuntime().exec(command, null, new File(this.jsAppDir))
       def bufferedReader = new BufferedReader(new InputStreamReader(cmdProcess.getInputStream()))
       def buff = ""
       def readBuffer = new StringBuffer()
@@ -164,23 +167,28 @@ class ReactNativeModules {
           readBuffer.append(buff)
       }
       reactNativeConfigOutput = readBuffer.toString()
-    } catch (Exception exception) {
-      this.logger.warn("${LOG_PREFIX}${exception.message}")
-      this.logger.warn("${LOG_PREFIX}Automatic import of native modules failed.")
 
-      def bufferedErrorReader = new BufferedReader(new InputStreamReader(cmdProcess.getErrorStream()))
-      def buff = ""
-      def readBuffer = new StringBuffer()
-      while ((buff = bufferedErrorReader.readLine()) != null){
-          readBuffer.append(buff)
+      if (!reactNativeConfigOutput) {
+        def bufferedErrorReader = new BufferedReader(new InputStreamReader(cmdProcess.getErrorStream()))
+        def errBuff = ""
+        def readErrorBuffer = new StringBuffer()
+        while ((errBuff = bufferedErrorReader.readLine()) != null){
+            readErrorBuffer.append(errBuff)
+        }
+        throw new Exception(readErrorBuffer.toString())
       }
-      this.logger.warn("${LOG_PREFIX}${readBuffer.toString()}")
-
-      return reactNativeModules
+    } catch (Exception exception) {
+      this.logger.error("${LOG_PREFIX}Automatic import of native modules failed.")
+      throw exception
     }
 
     def json = new JsonSlurper().parseText(reactNativeConfigOutput)
     def dependencies = json["dependencies"]
+    def project = json["project"]["android"]
+
+    if (project == null) {
+      throw new Exception("React Native CLI failed to determine Android project configuration. This is likely due to misconfiguration. Config output:\n${json.toMapString()}")
+    }
 
     dependencies.each { name, value ->
       def platformsConfig = value["platforms"];
@@ -211,7 +219,7 @@ class ReactNativeModules {
  *    Exported Extensions
  * ------------------------ */
 
-def autoModules = new ReactNativeModules(logger)
+def autoModules = new ReactNativeModules(logger, jsAppDir)
 
 ext.applyNativeModulesSettingsGradle = { DefaultSettings defaultSettings, String root = null ->
   if (root != null) {

--- a/packages/platform-android/native_modules.gradle
+++ b/packages/platform-android/native_modules.gradle
@@ -158,7 +158,9 @@ class ReactNativeModules {
     def reactNativeConfigOutput = ""
 
     try {
-      // When running Gradle from CWD that's outside of JS project, it's necessary to pass a CWD that's close to JS app,// so that npx can resolve to correct local CLI, instead of possibly installed global, corrupting config output
+      // Running npx from the directory of the JS app which holds this script in its node_modules.
+      // We do so, because Gradle may be ran from a different directory, that's outside of JS project,
+      // in which case npx wouldn't resolve correct `react-native` binary
       def cmdProcess = Runtime.getRuntime().exec(command, null, new File(this.jsAppDir))
       def bufferedReader = new BufferedReader(new InputStreamReader(cmdProcess.getInputStream()))
       def buff = ""

--- a/packages/platform-android/native_modules.gradle
+++ b/packages/platform-android/native_modules.gradle
@@ -160,7 +160,7 @@ class ReactNativeModules {
       }
       output = readBuffer.toString()
       if (!output) {
-        this.logger.error("${LOG_PREFIX}Unexpected empty result of running '${command}' command.")
+        this.logger.error("${LOG_PREFIX}Unexpected empty result of running '${command}' command from '${directory}' directory.")
         def bufferedErrorReader = new BufferedReader(new InputStreamReader(cmdProcess.getErrorStream()))
         def errBuff = ""
         def readErrorBuffer = new StringBuffer()
@@ -171,7 +171,7 @@ class ReactNativeModules {
       }
       return output
     } catch (Exception exception) {
-      this.logger.error("${LOG_PREFIX}Running '${command}' command failed.")
+      this.logger.error("${LOG_PREFIX}Running '${command}' command from '${directory}' directory failed.")
       throw exception
     }
   }
@@ -190,7 +190,13 @@ class ReactNativeModules {
     // in which case npx wouldn't resolve correct `react-native` binary
     def dir = new File(this.jsAppDir)
     def reactNativeConfigOutput = this.getCommandOutput(command, dir)
-    def json = new JsonSlurper().parseText(reactNativeConfigOutput)
+    def json
+    try {
+      json = new JsonSlurper().parseText(reactNativeConfigOutput)
+    } catch (Exception exception) {
+      this.logger.error("${LOG_PREFIX}Failed to parse React Native CLI configuration: ${exception.toString()}")
+      throw new Exception("Failed to parse React Native CLI configuration. Expected running '${command}' command from '${dir}' directory to output valid JSON, but it didn't. This may be caused by npx resolving to a legacy global react-native binary. Please make sure to uninstall any global 'react-native' binaries: 'npm uninstall -g react-native react-native-cli' and try again")
+    }
     def dependencies = json["dependencies"]
     def project = json["project"]["android"]
 

--- a/packages/platform-android/native_modules.gradle
+++ b/packages/platform-android/native_modules.gradle
@@ -145,45 +145,51 @@ class ReactNativeModules {
   }
 
   /**
-   * Runs a process to call the React Native CLI Config command and parses the output
-   *
-   * @return ArrayList < HashMap < String , String > >
+   * Runs a specified command using Runtime exec() in a specified directory.
+   * Throws when the command result is empty.
    */
-  ArrayList<HashMap<String, String>> getReactNativeConfig() {
-    if (this.reactNativeModules != null) return this.reactNativeModules
-    ArrayList<HashMap<String, String>> reactNativeModules = new ArrayList<HashMap<String, String>>()
-
-    def npx = Os.isFamily(Os.FAMILY_WINDOWS) ? "npx.cmd" : "npx"
-    def command = "${npx} --quiet --no-install react-native config"
-    def reactNativeConfigOutput = ""
-
+  String getCommandOutput(String command, File directory = null) {
     try {
-      // Running npx from the directory of the JS app which holds this script in its node_modules.
-      // We do so, because Gradle may be ran from a different directory, that's outside of JS project,
-      // in which case npx wouldn't resolve correct `react-native` binary
-      def cmdProcess = Runtime.getRuntime().exec(command, null, new File(this.jsAppDir))
+      def output = ""
+      def cmdProcess = Runtime.getRuntime().exec(command, null, directory)
       def bufferedReader = new BufferedReader(new InputStreamReader(cmdProcess.getInputStream()))
       def buff = ""
       def readBuffer = new StringBuffer()
-      while ((buff = bufferedReader.readLine()) != null){
-          readBuffer.append(buff)
+      while ((buff = bufferedReader.readLine()) != null) {
+        readBuffer.append(buff)
       }
-      reactNativeConfigOutput = readBuffer.toString()
-
-      if (!reactNativeConfigOutput) {
+      output = readBuffer.toString()
+      if (!output) {
+        this.logger.error("${LOG_PREFIX}Unexpected empty result of running '${command}' command.")
         def bufferedErrorReader = new BufferedReader(new InputStreamReader(cmdProcess.getErrorStream()))
         def errBuff = ""
         def readErrorBuffer = new StringBuffer()
-        while ((errBuff = bufferedErrorReader.readLine()) != null){
-            readErrorBuffer.append(errBuff)
+        while ((errBuff = bufferedErrorReader.readLine()) != null) {
+          readErrorBuffer.append(errBuff)
         }
         throw new Exception(readErrorBuffer.toString())
       }
+      return output
     } catch (Exception exception) {
-      this.logger.error("${LOG_PREFIX}Automatic import of native modules failed.")
+      this.logger.error("${LOG_PREFIX}Running '${command}' command failed.")
       throw exception
     }
+  }
 
+  /**
+   * Runs a process to call the React Native CLI Config command and parses the output
+   */
+  ArrayList<HashMap<String, String>> getReactNativeConfig() {
+    if (this.reactNativeModules != null) return this.reactNativeModules
+
+    ArrayList<HashMap<String, String>> reactNativeModules = new ArrayList<HashMap<String, String>>()
+    def npx = Os.isFamily(Os.FAMILY_WINDOWS) ? "npx.cmd" : "npx"
+    def command = "${npx} --quiet --no-install react-native config"
+    // Running npx from the directory of the JS app which holds this script in its node_modules.
+    // We do so, because Gradle may be ran from a different directory, that's outside of JS project,
+    // in which case npx wouldn't resolve correct `react-native` binary
+    def dir = new File(this.jsAppDir)
+    def reactNativeConfigOutput = this.getCommandOutput(command, dir)
     def json = new JsonSlurper().parseText(reactNativeConfigOutput)
     def dependencies = json["dependencies"]
     def project = json["project"]["android"]

--- a/packages/platform-android/native_modules.gradle
+++ b/packages/platform-android/native_modules.gradle
@@ -2,7 +2,7 @@ import groovy.json.JsonSlurper
 import org.gradle.initialization.DefaultSettings
 import org.apache.tools.ant.taskdefs.condition.Os
 
-def jsAppDir = buildscript.getProperties().get("sourceFile").toString().split("node_modules/@react-native-community/cli-platform-android/native_modules.gradle")[0]
+def jsAppDir = buildscript.sourceFile.toString().split("node_modules/@react-native-community/cli-platform-android/native_modules.gradle")[0]
 def generatedFileName = "PackageList.java"
 def generatedFilePackage = "com.facebook.react"
 def generatedFileContentsTemplate = """

--- a/packages/platform-android/native_modules.gradle
+++ b/packages/platform-android/native_modules.gradle
@@ -2,7 +2,7 @@ import groovy.json.JsonSlurper
 import org.gradle.initialization.DefaultSettings
 import org.apache.tools.ant.taskdefs.condition.Os
 
-def jsAppDir = buildscript.sourceFile.toString().split("node_modules/@react-native-community/cli-platform-android/native_modules.gradle")[0]
+def jsAppDir = buildscript.sourceFile.toString().split("node_modules/@react-native-community/cli-platform-android")[0]
 def generatedFileName = "PackageList.java"
 def generatedFilePackage = "com.facebook.react"
 def generatedFileContentsTemplate = """
@@ -185,9 +185,11 @@ class ReactNativeModules {
     ArrayList<HashMap<String, String>> reactNativeModules = new ArrayList<HashMap<String, String>>()
     def npx = Os.isFamily(Os.FAMILY_WINDOWS) ? "npx.cmd" : "npx"
     def command = "${npx} --quiet --no-install react-native config"
-    // Running npx from the directory of the JS app which holds this script in its node_modules.
-    // We do so, because Gradle may be ran from a different directory, that's outside of JS project,
-    // in which case npx wouldn't resolve correct `react-native` binary
+    /**
+     * Running npx from the directory of the JS app which holds this script in its node_modules.
+     * We do so, because Gradle may be ran with a different directory as CWD, that's outside of JS project
+     * (e.g. when running with -p flag), in which case npx wouldn't resolve correct `react-native` binary.
+     */
     def dir = new File(this.jsAppDir)
     def reactNativeConfigOutput = this.getCommandOutput(command, dir)
     def json

